src/server/identity/server/provisioner.go

```diff
@@ -0,0 +1,186 @@
+package server
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net/url"
+
+	det "github.com/determined-ai/determined/proto/pkg/apiv1"
+	userv1 "github.com/determined-ai/determined/proto/pkg/userv1"
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	mlc "github.com/pachyderm/pachyderm/v2/src/internal/middleware/logging/client"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/metadata"
+)
+
+type provisioner interface {
+	FindUser(ctx context.Context, name string) (*User, error)
+	CreateUser(context.Context, *User) (*User, error)
+	FindGroup(ctx context.Context, name string) (*Group, error)
+	CreateGroup(context.Context, *Group) (*Group, error)
+	SetUserGroups(context.Context, *User, []*Group) error
+	Close() error
+}
+
+type User struct {
+	id         int32
+	name       string
+	prevGroups []*Group
+}
+
+type Group struct {
+	id   int32
+	name string
+}
+
+type determinedProvisioner struct {
+	conn  *grpc.ClientConn
+	dc    det.DeterminedClient
+	token string
+}
+
+type DeterminedConfig struct {
+	MasterURL string
+	Username  string
+	Password  string
+	TLS       bool
+}
+
+type errNotFound struct{}
+
+func (e errNotFound) Error() string {
+	return "not found"
+}
+
+func withDetToken(ctx context.Context, token string) context.Context {
+	return metadata.AppendToOutgoingContext(ctx, "x-user-token", fmt.Sprintf("Bearer %s", token))
+}
+
+func NewDeterminedProvisioner(ctx context.Context, config DeterminedConfig) (provisioner, error) {
+	tlsOpt := grpc.WithTransportCredentials(insecure.NewCredentials())
+	if config.TLS {
+		tlsOpt = grpc.WithTransportCredentials(credentials.NewTLS(&tls.Config{
+			InsecureSkipVerify: true,
+		}))
+	}
+	determinedURL, err := url.Parse(config.MasterURL)
+	if err != nil {
+		return nil, errors.Wrapf(err, "parsing determined url %q", config.MasterURL)
+	}
+	conn, err := grpc.DialContext(ctx, determinedURL.Host, tlsOpt, grpc.WithStreamInterceptor(mlc.LogStream), grpc.WithUnaryInterceptor(mlc.LogUnary))
+	if err != nil {
+		return nil, errors.Wrapf(err, "dialing determined at %q", determinedURL.Host)
+	}
+	defer conn.Close()
+	dc := det.NewDeterminedClient(conn)
```

Comment:
This captures `conn`, which will be closed when `NewDeterminedProvisioner` returns.